worker/remoterelations/remoterelations.go

```diff
@@ -240,23 +252,39 @@ func (w *Worker) handleApplicationChanges(applicationIds []string) error {
 
 	for i, result := range results {
 		name := applicationIds[i]
-		if result.Error != nil {
-			// The the remote application has been removed, stop its worker.
-			if params.IsCodeNotFound(result.Error) {
-				if err := w.runner.StopWorker(name); err != nil {
-					return err
-				}
+
+		// The remote application may refer to an offer that has been removed from
+		// the offering model, or it may refer to a new offer with a different UUID.
+		// If it is for a new offer, we need to stop any current worker for the old offer.
+		offerChanged := false
+		var remoteApp *params.RemoteApplication
+		if result.Error != nil && !params.IsCodeNotFound(result.Error) {
```

Comment:
to make this clearer I'd go
```
appGone := params.IsCodeNotFound(err) || errors.IsNotFound(err)
if result.Error != nil && !appGone {
    return errors.Annotatef(result.Error, "querying remote application %q", name)
}
```

although it is annoying that params package is leaking to here and the error handling isn't dealt with at the api package

Reply:
IsCodeNotFound(err) will panic if err is nil, so it needs the nil check

Reply:
i tweaked it a bit